### ISSUE TITLE
FlashAttention parity: generic T, 3D, bias, offset, block-size config, attention weights, backward, CPU fallback

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet.Tensors#198]
## Context

`AiDotNet.Tensors` 0.46.0 exposes `RecordingGpuBackend.FlashAttention` and `RecordingGpuBackend.ScaledDotProductAttention` as GPU-only, float32-only, 4D-only fused kernels. Consumers (`AiDotNet` most notably) maintain their own full-featured CPU FlashAttention under `src/NeuralNetworks/Attention/FlashAttention.cs` — ~900 LOC of online-softmax block-tiled attention with features the Tensors kernel lacks. For full parity at the Tensors level, we want Tensors to cover every capability.

## Current Tensors signature

\`\`\`csharp
void FlashAttention(
    IGpuBuffer query, IGpuBuffer key, IGpuBuffer value,
    IGpuBuffer output,
    IGpuBuffer mask,
    int batch, int numHeads, int seqLen, int headDim,
    float scale, bool isCausal);

void ScaledDotProductAttention(
    IGpuBuffer query, IGpuBuffer key, IGpuBuffer value,
    IGpuBuffer output, IGpuBuffer attentionWeights, IGpuBuffer mask,
    int batch, int numHeads, int seqLen, int headDim,
    float scale, bool isCausal);
\`\`\`

### A. Generic `T`
Both kernels are `float`-only (`IGpuBuffer` holds `Single[]`). Need generic versions for `double`, `Half`, `bfloat16`, and the FP8 types once issue #197 lands:

\`\`\`csharp
public static class FusedAttention<T>
{
    public static (Tensor<T> Output, Tensor<T>? AttentionWeights) Forward(
        Tensor<T> query, Tensor<T> key, Tensor<T> value,
        FlashAttentionConfig config,
        Tensor<T>? attentionBias = null,
        IEngine? engine = null);

    public static (Tensor<T> GradQuery, Tensor<T> GradKey, Tensor<T> GradValue) Backward(
        Tensor<T> gradOutput,
        Tensor<T> query, Tensor<T> key, Tensor<T> value,
        Tensor<T> output, Tensor<T> logsumexp,   // stats from forward
        FlashAttentionConfig config,
        Tensor<T>? attentionBias);
}
\`\`\`

Place under `AiDotNet.Tensors.Engines.Autodiff` (it fits with existing tape-autodiff ops).

### B. 3D tensor support

Current GPU kernel requires rank-4 `[batch, heads, seqLen, headDim]`. Need rank-3 acceptance `[batch, seqLen, headDim]` treated as `numHeads=1`. Auto-dispatch based on `Shape.Length`.

### C. `attentionBias` parameter

Additive bias tensor broadcast over attention scores before softmax. Required for:
- ALiBi positional bias
- Relative position encodings (T5-style)
- Custom masking beyond on/off causal/padding masks

Signature: `Tensor<T>? attentionBias` — either `[batch, heads, seqLenQ, seqLenKV]` or broadcastable.

### D. `queryOffset` / KV-cache support

For autoregressive inference and sliding-window attention, the query can be a slice of the full sequence while K/V represent the full history. Current signature conflates `seqLenQ == seqLenKV`. Needed:

\`\`\`csharp
int queryOffset = 0,   // where Q starts within the KV sequence
int seqLenQ,           // possibly < seqLenKV
int seqLenKV
\`\`\`

Causal mask honors `queryOffset` so token `q_i` attends to all K positions `k <= queryOffset + i`.

### E. Configurable block sizes

`FlashAttentionConfig` exists in AiDotNet with:

\`\`\`csharp
public class FlashAttentionConfig
{
    public int BlockSizeQ { get; set; } = 64;
    public int BlockSizeKV { get; set; } = 64;
    public double? ScaleFactor { get; set; } = null;       // null = 1/sqrt(headDim)
    public bool IsCausal { get; set; } = false;
    public bool ReturnAttentionWeights { get; set; } = false;
    public double? DropoutRate { get; set; } = null;
    public static FlashAttentionConfig Default { get; }
}
\`\`\`

Port to Tensors. The GPU kernel today hard-codes block sizes; surface them.

### F. Optional attention-weights output

For debugging, interpretability, and visualization. Off by default (pure FlashAttention streams through tiles and discards weights). When enabled, materialize the `[batch, heads, seqLenQ, seqLenKV]` weights tensor. Slower but crucial for researchers.

### G. Backward pass

Current kernel is forward-only. Need gradient computation suitable for autodiff:
- Recompute attention weights from forward's saved `logsumexp` + inputs
- Compute `dQ`, `dK`, `dV` in tiles (standard FlashAttention-2 backward)
- Optional: gradient wrt `attentionBias` if enabled

### H. CPU fallback

GPU kernel requires `IDirectGpuBackend`. Need CPU implementation for:
- Environments without GPU
- Debugging (numerically-compare forward output against reference SDPA)
- ONNX export / deterministic CI

AiDotNet's implementation in `src/NeuralNetworks/Attention/FlashAttention.cs` is Apache-2.0 and is a ready CPU port. Place under `AiDotNet.Tensors.Engines.Autodiff.FusedAttention<T>`; CpuEngine provides the implementation.

### I. `ScaledDotProductAttention` — same gap list

Same generic-T, 3D support, queryOffset, bias, and backward pass requirements apply. Unfused SDPA is the numerical reference for the fused FlashAttention output, so they must produce bitwise-identical results on CPU given the same inputs.

## Reference implementation

Full source at \`ooples/AiDotNet\` \`src/NeuralNetworks/Attention/FlashAttention.cs\` — Apache-2.0, ~900 LOC. Implements forward (3D + 4D) with attentionBias + queryOffset + block config + optional attention weights; backward is at line 561.

\`FlashAttentionConfig.cs\` in \`src/NeuralNetworks/Attention/FlashAttentionConfig.cs\`.

## Acceptance tests

- Forward output matches unfused \`(QK^T / sqrt(d)) * softmax * V\` numerically on CPU to within 1e-5 for \`float\`, 1e-9 for \`double\`, 1e-3 for \`Half\`
- CPU and GPU backends produce identical results (within precision floor)
- Causal mask with \`queryOffset > 0\` matches reference KV-cache SDPA
- Backward gradients verified by finite-difference against small inputs
- attentionBias passes ALiBi and T5 relative-position test vectors
- Memory: forward is O(seqLen) in intermediate allocations, not O(seqLen²), via online softmax

## Why this matters

PyTorch's \`torch.nn.functional.scaled_dot_product_attention\` auto-dispatches to FlashAttention, mem-efficient, or math kernels based on inputs and supports bias, causal, and arbitrary T. Without bringing Tensors up to that surface, consumers will keep duplicating attention code. This closes the parity.

## Relationship with issue #197

Independent. Can ship as \`0.47.0\` alongside or separately. If FP8 ships first, this gets \`Float8E4M3\`/\`E5M2\` generic dispatch for free.

## Estimated scope

~1,500 LOC implementation (CPU + GPU parity), ~600 LOC tests. Full-fat FlashAttention-2 backward can reuse existing \`RecordingGpuBackend\` op set.
